### DISCUSSION TITLE
v4.0.0

### DISCUSSION THREAD

**Discussion by devlead** [cake-build/cake#4270] [Announcements]
As part of this release we had [15 issues](https://github.com/cake-build/cake/milestone/89?closed=1) closed.

__Breaking changes__

- [__#4132__](https://github.com/cake-build/cake/issues/4132) Add File APIs for setting timestamps (creation time, last write time, last access time)
- [__#4266__](https://github.com/cake-build/cake/issues/4266) Update LatestPotentialBreakingChange to 4.0.0

__Bugs__

- [__#4150__](https://github.com/cake-build/cake/issues/4150) Cake script is contributing unactionable diagnostics in VS Code Problems pane
- [__#4197__](https://github.com/cake-build/cake/issues/4197) Execution of Cake script fails if an addin defines an alias that uses nullable reference types in its signature

__Feature__

- [__#4250__](https://github.com/cake-build/cake/issues/4250) Update System.Collections.Immutable to 8.0.0

__Improvements__

- [__#4249__](https://github.com/cake-build/cake/issues/4249) Update Microsoft.Extensions.DependencyInjection to 8.0.0
- [__#4251__](https://github.com/cake-build/cake/issues/4251) Update System.Reflection.Metadata 8.0.0
- [__#4260__](https://github.com/cake-build/cake/issues/4260) Unzip alias should support overwrite files

<hr /><em>This discussion was created from the release <a href='https://github.com/cake-build/cake/releases/tag/v4.0.0'>v4.0.0</a>.</em>

**Comment by MihaMarkic** [comment]
Is this the right place to discuss issues related to this version?

I think Frosting broke something when invoked from powershell the old way
```powershell
$cakeArguments = @("$Script");
dotnet run --project build/Build.csproj -- $cakeArguments
```
will output an error:
```
Error: Unknown command ''.
        No such command
```

Same stuff works fine with 3.0.0.

**Reply by devlead** [reply]
Do you have the same error with `3.2.0` or does that work fine?

Which version of the .NET SDK are you using?

**Reply by MihaMarkic** [reply]
3.2.0 -> same error, using .NET 8

**Reply by MihaMarkic** [reply]
Also using Powershell Core 7.4.0 if that matters.

**Reply by ltrzesniewski** [reply]
I had this error (`Unknown command ''`) on Linux. I was using an older bootstrapper script which had the following code block:

https://github.com/cake-build/resources/blob/416f110bc1baf5b7a86598374e3b9cfca4c2a634/dotnet-framework/build.sh#L37-L45

Turns out that enumerating `"$@"` with `for` and using `shift` don't work well together. Replacing `for i in "$@"; do` with `while [[ $# -gt 0 ]]; do` solved the issue.

**Reply by pascalberger** [reply]
> I had this error (`Unknown command ''`) on Linux. I was using an older bootstrapper script which had the following code block:

Please note that this is an example bootstrapper for the deprecated [.NET Framework runner](https://cakebuild.net/docs/running-builds/runners/deprecated/cake-runner-for-dotnet-framework), which is no longer supported since Cake 2.0. See [Cake .NET Tool](https://cakebuild.net/docs/running-builds/runners/dotnet-tool) or [Cake Frosting](https://cakebuild.net/docs/running-builds/runners/cake-frosting) for example bootstrappers for supported runners of Cake 4.0.

**Reply by ltrzesniewski** [reply]
Yes, I'm aware of that. But we use Cake for a long time, and the bootstrapper we currently use evolved from that one. I can assure you we use the Cake v4 dotnet tool now. 🙂

We had to keep some of the old bootrapper logic though for several reasons, one of them being that our build server is configured to pass a `--script` argument for hundreds of projects, and it was just easier that way. But our current bootstrapper is much more similar to the dotnet-tool one than to the legacy one from dotnet-framework.
